File: Java-Visualizer/jsonFilter/JsonFilter.java
package jsonfilter;
 
import java.io.*;
import java.util.Stack;
import java.lang.Object; 
import java.util.Scanner; 
import java.util.Arrays;
import java.util.ArrayList; 
import java.lang.String;
import java.nio.ByteBuffer;


/**
 *
 * @author Kyle Reinholt 
 * 
 * @purpose This program will read in a json text file, filter out the execution points that are between the function startTraceNow() and endTraceNow(), 
 *          and output the filtered executionPoints to a javaScript page that will be read in by the html file for display. 
 * 
 *
 */

class Event { //This class makes it easier to analyze an executionPoint.  
    
    private int lineNumber; 
    private String trace; 
    
    Event() { 
        
        lineNumber = 0; 
        trace = ""; 
    }
    
    public void setEvent(String exePoint){
        
        trace = exePoint; 
    }
    
    public void setLineNumber(int line) { 
        
        lineNumber = line; 
    }
    
    public String getEvent() { 
        
       return trace; 
    }
    
    public int getLineNumber() { 
        
        return lineNumber; 
    }
    
}

class EventManager { //This class provides a way to weed out executionPoints that aren't intended to be visualized. 
                     // the ArrayList listOfEvents contains all of the executionPoints between startTraceNow(); and endTraceNow();
                     // the ArrayList filteredEvents 'should' only contain executionPoints that are intended for visualization. 
    
    private ArrayList<Event> listOfEvents;
    private ArrayList<Event> filteredEvents; 
    
    EventManager() {
        
        listOfEvents = new ArrayList<>();
        filteredEvents = new ArrayList<>(); 
    }
    
    public int getNumOfEvents() {
        
        return filteredEvents.size(); 
    }
    
    public Event getEvent(int index){
        
        return filteredEvents.get(index);  
    }
    
    public void addEvent(Event newEvent) { 
        
        listOfEvents.add(newEvent); 
    }
    
    public void modifyLineNums(ArrayList<String> userCode) {
        
        int lineNumber = 0; 
        int eventNumber = 0; 
        
        while(lineNumber != userCode.size()) {
            
           Event modify = new Event(); 
           String tempString = ""; 
           int tempLine = 0; 
           
           modify = filteredEvents.get(eventNumber); 
           
           tempString = modify.getEvent(); 
           
           tempLine = modify.getLineNumber(); 
           
           if(userCode.get(lineNumber) == "newline"){
               
               
               lineNumber = lineNumber + 1; 
           }
           else{
                
                String originalLine = Integer.toString(tempLine);
                String newLine = Integer.toString(lineNumber+1);

                tempString = tempString.replace(originalLine, newLine);

                filteredEvents.get(eventNumber).setEvent(tempString);
                filteredEvents.get(eventNumber).setLineNumber(lineNumber+1);

                eventNumber = eventNumber + 1; 
                lineNumber = lineNumber + 1; 
           }
           
        }
        
        
        
        int ogLine = filteredEvents.get(eventNumber).getLineNumber();
        String newTrace = filteredEvents.get(eventNumber).getEvent(); 
               
        String oldLine = Integer.toString(ogLine);
        String newLine = Integer.toString(filteredEvents.get(eventNumber-1).getLineNumber());

        newTrace = newTrace.replace(oldLine, newLine);

        filteredEvents.get(eventNumber).setEvent(newTrace); 
    }
    
/* *** Start of Test Functions *** /*
    public void printAllLines(){a
        
        for(int i = 0; i < listOfEvents.size(); i++) {
            
            
            System.out.print("Line number: "); 
            System.out.print(listOfEvents.get(i).getLineNumber()); 
            System.out.println();  
        }
    }
    
    public void printFilteredEvents(){ 
        
        for(int l = 0; l < filteredEvents.size(); l++) {
           
            System.out.print("Line number: "); 
            System.out.print(filteredEvents.get(l).getLineNumber()); 
            System.out.println();  
        }
    }
/* *** End of Test Functions *** */     
    
    public void verifyEvents() { 
        
        int currentLine = 0;
        int nextLine = 0; 
        int eventCounter = 0; 
        
        int originalLine = listOfEvents.get(0).getLineNumber(); 
        
        currentLine = listOfEvents.get(0).getLineNumber(); 
        
        for(int k = 1; k < listOfEvents.size(); k++) {
            
            nextLine = listOfEvents.get(k).getLineNumber(); 
            
            if(currentLine+1 ==  nextLine && nextLine > originalLine || currentLine+2 ==  nextLine && nextLine > originalLine || currentLine+3 ==  nextLine && nextLine > originalLine  || currentLine+4 ==  nextLine && nextLine > originalLine || currentLine+5 ==  nextLine && nextLine > originalLine  || currentLine+6 ==  nextLine && nextLine > originalLine ){
                
                filteredEvents.add(listOfEvents.get(eventCounter));
                eventCounter = eventCounter + 1; 
                currentLine = nextLine;
  
            }
            else { 
                
                currentLine = nextLine;
                eventCounter = eventCounter + 1; 
                
            }
            
            
        }
        
        filteredEvents.add(listOfEvents.get(listOfEvents.size()-1)); 
       
    }
    
}

class CodeAnalyzer { 
    
    private String codeTrace; 
    private ArrayList<String> codeList;
    
    CodeAnalyzer(String jsonCode) {
        
        codeList = new ArrayList<>(); 
        setCodeTrace(jsonCode);  
    }
    
    public ArrayList getCodeList() {
        
        return codeList; 
    }
    
    public void setCodeTrace(String json){
        
        codeTrace = json; 
    }
    
    public void isolateStudentCode(){
        
        String modify = codeTrace; 
        
        String[] isolateMain = modify.split("main"); //retrieves everything after main function. We are only conserned with code that is being executed within main.  
        
        String[] isolateStart = isolateMain[1].split("startTraceNow\\(\\)\\;"); //This will grab everything after startTrace
        
        String[] isolateEnd = isolateStart[1].split("endTraceNow\\(\\)\\;");
        
        String myCode = isolateEnd[0];
        
        String newLine = myCode.replace("\\n", "newline"); //This was the only way I could get it to split at each newline.   
        
        String myString = newLine.replaceFirst("newline", ""); 
        
        myString = myString.trim(); 
        
        String[] studentCode = myString.split("newline"); 
        
        for(int o = 0; o < studentCode.length; o++){
             
            String temp = studentCode[o].replaceAll("\\s+",""); 
            
            if(temp.isEmpty() == true || temp == " " || temp == ""){
                
                studentCode[o] = "newline"; 
                codeList.add(studentCode[o]); 
            }
            else if(temp.length() < 3) { //This will capture any case where there is a blank line. Blank lines must be
                                                   //added to the codeList so when the filteredEvents' line numbers are being modified
                                                   //the event will not get assigned a line number that corresponds to a blank line. 
                                                   //Although everything the visualizer needs is in the execution point, OPT will not 
                                                   //visualize an execution point if the line is blank.
                
                studentCode[o] = "newline"; 
                codeList.add(studentCode[o]);
            }
            else{
                
                codeList.add(studentCode[o]);
            }
             
        } 
        
    }
    
    public void removeUnwantedCode() {
        
        for(int j = 0; j < codeList.size(); j++){
            
            if(codeList.get(j) == "newline"){
                
                codeList.set(j,""); 
            }
            else {
                
                //Do Nothing 
            }
        }
    }
    
}

class TraceAnalyzer {
    
    private String originalTrace; 
    private Stack<Character> symbolStack;
    private ArrayList<String> exePointList;  
    private int executionPoints; //integer counter of how many execution points there are
    private UniqueFileWriter jsonWriter;
    private EventManager eventManager; 
    
    TraceAnalyzer(String jsonTrace) {
        
        originalTrace = jsonTrace;
        exePointList = new ArrayList<>();
        eventManager = new EventManager(); 
        executionPoints = 0; 
    }
    
    public void filterExePoints() { 
        
        String copyTrace = originalTrace; 
        String executionPoint = ""; 
        
        boolean startSwitch = false; 
        boolean endSwitch = false;
        
        symbolStack = new Stack<>();
        
        for (int index = 0; index < copyTrace.length(); index++) {
            
            char currentSymbol = copyTrace.charAt(index);
            
            executionPoint = executionPoint + currentSymbol;  
            
            switch (currentSymbol) { //This switch statement isolates execution points by pairing up braces. 
            case '(':
            case '[':
            case '{': 
                symbolStack.push(currentSymbol);
                break;

            case ')':
            case ']':
            case '}':
                if (!symbolStack.isEmpty()) {
                    char symbolStackTop = symbolStack.pop();
                    if ((currentSymbol == '}' && symbolStackTop != '{')
                            || (currentSymbol == ')' && symbolStackTop != '(')
                            || (currentSymbol == ']' && symbolStackTop != '[')) {
                        
                        return;
                        }
                } 
                else{
                    
                    return;                    
                }
                break;
            case ',': 
                 if ( symbolStack.size() == 0){
                     
                     executionPoint = executionPoint.substring(0, executionPoint.length()-1); //This removes the trailing comma at the end of each exePoint
                     
                     if( executionPoint.contains("startTraceNow")){
                         
                         startSwitch = true;  
                         executionPoint = "";
                     }
                     else if( executionPoint.contains("endTraceNow")){
                         
                         endSwitch = true;
                         break; 
                     }
                     else {
                         
                         if(verifyExePoint(startSwitch,endSwitch,executionPoint)){ //If start switch is true, and endSwitch is false, the executionPoint will be made into an event
                             
                             executionPoints = executionPoints + 1; //integer counter of how many execution points there are
                             executionPoint = ""; 
                         }
                         else {
                             
                             executionPoint = ""; 
                         }
                      
                     }
                                                           
                 }
                 else{
                     //Do Nothing, this may be a bad practice 
                 }
                 break; 
                 
            default:
                break;
                
            }
        
        }
      
    }
    
    public void handleEvents(ArrayList<String> codeList) { 
        
       eventManager.verifyEvents();
       eventManager.modifyLineNums(codeList);
    }
    
    public void outputFilteredJSON(ArrayList<String> inList) {
        
        jsonWriter = new UniqueFileWriter();
        
        jsonWriter.writeToFile("var testvisualizerTrace = ");
        
        jsonWriter.writeToFile("{" + '"' + "code" + '"' + ':' + '"');
        for(int p = 0; p < inList.size(); p++) {
            
            jsonWriter.writeToFile(inList.get(p));
            jsonWriter.writeToFile("\\n");
        }
        
        jsonWriter.writeToFile("" + '"' + ',' + '"' + "trace" + '"' + ':' + '[');
        
        for( int m = 0; m < eventManager.getNumOfEvents(); m++){
            
            Event myEvent = new Event();  
            myEvent = eventManager.getEvent(m);
            
            if( m == eventManager.getNumOfEvents() - 1)
            {
                jsonWriter.writeToFile(myEvent.getEvent());
                jsonWriter.writeToFile("]" + ',' + '"' + "userlog" + '"' + ":" + '"' + "Debugger VM maxMemory: 807M" + "\\n" + '"' + "}");
                jsonWriter.writeToFile("\n");
                jsonWriter.writeToFile("$(document).ready(function() { \n \n \t var testvisualizer = new ExecutionVisualizer('testvisualizerDiv', testvisualizerTrace,{embeddedMode: false, lang: 'java', heightChangeCallback: redrawAllVisualizerArrows}); \n \n \tfunction redrawAllVisualizerArrows() { \n \n \t \t if (testvisualizer) testvisualizer.redrawConnectors(); \n \t } \n \n $(window).resize(redrawAllVisualizerArrows); \n});");
            }
            else 
            {
                jsonWriter.writeToFile(myEvent.getEvent());
                jsonWriter.writeToFile(",");
                jsonWriter.writeToFile("\n");
                jsonWriter.flushFile();
            }
           
        }
        
        jsonWriter.closeFile();
    
    }
    
    
    /* *** Start of Test Functions *** /* 
    public void printExecutionTraces(){
        
        //Test function
        
        for(int i = 0; i < exePointList.size(); i++){
            
            System.out.println(exePointList.get(i)); 
        }
    } 
    
    public void numOfExePnts(){
        
        System.out.print("There are: "); 
        System.out.print( executionPoints ); 
        System.out.print(" Execution Points."); 
    }
    /* *** End of Test Functions *** */ 
    
    public boolean verifyExePoint(boolean on, boolean off, String inPoint) {
        
        boolean addExePoint = false; 
        Event exeTrace = new Event(); 
        
        if(on == true && off == false) //If the exePoint makes it here, it qualifies as an event, however, the event must then be verified by the event manager
        {
            exePointList.add(inPoint);  
            exeTrace.setEvent(inPoint);
            exeTrace.setLineNumber(extractLineNum(inPoint)); 
            eventManager.addEvent(exeTrace);
            addExePoint = true; 
        }
        else if(on == false && off == false) 
        {
            addExePoint = false; 
        }
        else 
        {
            addExePoint = false; 
        }
        
        return addExePoint;
    }
    
    public int extractLineNum(String input){
        
        int lineNumber = 0; 
        
        lineNumber = new Scanner(input).useDelimiter("\\D+").nextInt();
        
        return lineNumber;
    }
    
}

class UniqueFileWriter{
    
    private File filteredTrace = new File("filteredJSON.js"); 
    private FileWriter fw; 
    
    UniqueFileWriter() {
        
       try{
        
           fw = new FileWriter(filteredTrace);
           
       }catch (IOException iox) {
            //do stuff with exception
            iox.printStackTrace();
       } 
    }
    
    public void flushFile() { 
        
       try {  
        
           fw.flush();
           
       }catch (IOException iox) { 
           
           iox.printStackTrace();
       }
    }
    
    public void closeFile() { 
        
       try{
            
           fw.close();
           
       }catch (IOException iox) {
        
           iox.printStackTrace(); 
       }
    }
    
    public void writeToFile(String input) { 
        
        
       try{
            
           fw.write(input);
           
       }catch (IOException iox) {
        
           iox.printStackTrace(); 
       }
    }
}

class UniqueFileReader{ 
    
    private int lineNumber; 
    private CodeAnalyzer analyzeCode; 
    private TraceAnalyzer analyzeTrace; 
    
    UniqueFileReader(String inputFile) {
       
        lineNumber = 0; 
        
        readFile(inputFile); 
    }
    
    public void readFile(String fileName) {     

        String jsonFile = fileName;

        String line = null; 
        String lineCopy = null; 

        try {
            // FileReader reads text files in the default encoding.
            FileReader fileReader = 
                new FileReader(jsonFile);

            // Always wrap FileReader in BufferedReader.
            BufferedReader bufferedReader = 
                new BufferedReader(fileReader);

            while((line = bufferedReader.readLine()) != null) {
                
                lineCopy = line; 
                
                String splitter = "" + '"' + ',' + '"';
                String traceSplitter = "" + '"' + "trace" + '"' + ':'; 
                
                String[] jsonCode = line.split(splitter);
                String[] jsonTrace = lineCopy.split(traceSplitter);
                
                splitter = "" + '"' + ':' + '"'; 
                
                String[] codeSection = jsonCode[0].split(splitter); 
                
                String codeSegment = codeSection[1]; 
                String traceSegment = jsonTrace[1]; 
                
                analyzeCode = new CodeAnalyzer(codeSegment);
                analyzeTrace = new TraceAnalyzer(traceSegment.substring(1));
                
                analyzeTrace.filterExePoints(); // Grabs execution points between startTraceNow() and endTraceNow() from backend trace string 
                 
                analyzeCode.isolateStudentCode();
                
                analyzeTrace.handleEvents(analyzeCode.getCodeList()); //Events cannot be filtered without knowing which line they correspond with.
                                                                      // The user code is sent to the eventVerifier to ensure the execution point
                                                                      // matches with the line of code that created it. 
                analyzeCode.removeUnwantedCode(); 
               
                analyzeTrace.outputFilteredJSON(analyzeCode.getCodeList()); //passing in the code trace to be printed to the file                                                                             
            }                                                               // before the execution points. 

            bufferedReader.close();            
        }
        catch(FileNotFoundException ex) {
            
            System.out.println(
                "Unable to open file '" + 
                jsonFile + "'");                
        }
        catch(IOException ex) {
            
            System.out.println(
                "Error reading file '" 
                + jsonFile + "'");                  
        }

    }

}

public class JsonFilter {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        
        UniqueFileReader myFileReader = new UniqueFileReader("jsonfile.txt");
   
    }
    
}

File: Java-Visualizer/jsonFilter/JsonFilter.py
#Author: Kyle Reinholt
#Date: 6/25/15
#Program: Json Filter
#Purpose: To extract execution points from the json trace produced from InMemory.java that are intended for visualization

import sys

with open('jsontrace.txt','r') as f:
    read_data = f.read()

f.closed

splitter = '"' + "trace" + '"' + ':'

userCode, wholeTrace = read_data.split(splitter)

class Event(object):
    
    def __init__(self):
        self.trace = ""
        self.lineNumber = 0

    def set_Event(self, trace):
        self.trace = trace

    def set_Line(self, lineNumber):
        self.lineNumber = lineNumber
    
    def get_Event():
        return self.trace

    def get_Line():
        return self.lineNumber

    
class EventManager(object):
    
    def __init__(self):
        self.listOfEvents = []
        self.filteredEvents = []

    def get_Line_Number(self, index):
        if is_empty(self.listOfEvents) == True:
            print("list is empty")
        else:
            tempEvent = self.listOfEvents[index]
            return tempEvent.lineNumber

    def set_Event(self, index, event):
        self.filteredEvents[index] = event
    
    def get_Event(index):
        return self.filteredEvents[index]

    def add_Event(self, event):
        self.listOfEvents.append(event)

    def trace_List(self):
        myList = []
        for x in range(0,len(self.filteredEvents)):
            temp = Event()
            temp = self.filteredEvents[x]
            myList.append(temp.trace)

        return myList

    def print_Events(self):
        if (len(self.filteredEvents) == 0):
            print("List of events is empty")
        else:
            for x in range(0, len(self.filteredEvents)):
                tempEvent = self.filteredEvents[x]
                print(tempEvent.trace)

    def verify_Events(self):
        currentLine = 0
        nextLine = 0
        eventCounter = 0
        counter = 1
        
        originalLineNum = self.get_Line_Number(0)
        currentLine = originalLineNum

        length = len(self.listOfEvents) - 1
    
        while counter < length+1:
            nextLine = self.get_Line_Number(counter)

            if (currentLine+1 == nextLine and nextLine > originalLineNum) or (currentLine+2 == nextLine and nextLine > originalLineNum) or (currentLine+3 == nextLine and nextLine > originalLineNum) or (currentLine+4 == nextLine and nextLine > originalLineNum) or (currentLine+5 == nextLine and nextLine > originalLineNum):
                self.filteredEvents.append(self.listOfEvents[eventCounter])
                eventCounter = eventCounter + 1
                currentLine = nextLine
                counter = counter + 1
            else:
                currentLine = nextLine
                eventCounter = eventCounter + 1
                counter = counter + 1

        
        self.filteredEvents.append(self.listOfEvents[length])

    def modify_Lines(self, code):
        lineNumber = 0
        eventNumber = 0

        while lineNumber != len(code):
            modify = Event()
            tempString = ""
            tempLine = 0

            modify = self.filteredEvents[eventNumber]
            tempString = modify.trace
            tempLine = modify.lineNumber

            if(code[lineNumber] == "newline"):
                lineNumber = lineNumber + 1
            else:
                originalLine = str(tempLine)
                newLine = str(lineNumber+1)
                tempString = tempString.replace(originalLine,newLine)
                modifiedEvent = Event()
                modifiedEvent.set_Event(tempString)
                modifiedEvent.set_Line(lineNumber+1)
                self.filteredEvents[eventNumber] = modifiedEvent

                eventNumber = eventNumber + 1
                lineNumber = lineNumber + 1

        lastEvent = Event()
        lastEvent = self.filteredEvents[eventNumber]
        tempLine = lastEvent.lineNumber
        tempString = lastEvent.trace

        oldLine = str(tempLine)

        secondTLEvent = Event()
        secondTLEvent = self.filteredEvents[eventNumber-1]
        otherLine = secondTLEvent.lineNumber

        newLine = str(otherLine)

        tempString = tempString.replace(oldLine,newLine)
        
        modifiedEvent = Event()
        modifiedEvent.set_Event(tempString)
        modifiedEvent.set_Line(otherLine)
        self.filteredEvents[eventNumber] = modifiedEvent        
                
                
    
class TraceAnalyzer(object):

    def __init__(self): 
        self.eventManager = EventManager()

    def handleEverything(self, userCode, inTrace):
        filterPoints = self.exe_Point_Finder
        filterPoints(inTrace)
        
        filteredEvents = self.eventManager.verify_Events
        filteredEvents()
        modifyLines = self.eventManager.modify_Lines(userCode) 

        events = []
        events = self.eventManager.trace_List()

        orig_stdout = sys.stdout
        f = file('filteredJSON.js', 'w')
        sys.stdout = f

        first = "var testvisualizerTrace = {\"code\":\""

        code = ""
        
        for x in range (0,len(userCode)):
            if userCode[x] == "newline":
                code = code + "\\n" + " "
            else:
                code = code + userCode[x] + "\\n"
        second = "\",\"trace\":["

        trace = ""
        for y in range(0,len(events)):
            if y == len(events)-1:
                tempString = events[y]
                tempString = tempString[:-1]
                trace = trace + tempString
                trace = trace + "],\"userlog\":\"Debugger VM maxMemory: 807M \\n \"}"
                trace = trace + "\n" + "$(document).ready(function() { \n \n \t var testvisualizer = new ExecutionVisualizer('testvisualizerDiv', testvisualizerTrace,{embeddedMode: false, lang: 'java', heightChangeCallback: redrawAllVisualizerArrows}); \n \n \tfunction redrawAllVisualizerArrows() { \n \n \t \t if (testvisualizer) testvisualizer.redrawConnectors(); \n \t } \n \n $(window).resize(redrawAllVisualizerArrows); \n});"
            else: 
                tempString = events[y]
                trace = trace + tempString + "\n"
                
        printToFile = first + code + second + trace

        print printToFile
        
        sys.stdout = orig_stdout
        f.close()
        
    def is_empty(self, any_structure):
        if any_structure:
            return False
        else:
            return True    

    def extract_Line_Num(self, string):

        line = string.replace('"'," ")
        line = line.replace('{'," ")
        line = line.replace(':'," ")
        line = line.replace(','," ")
        line = line.replace('['," ")
        line = line.replace('('," ")
        line = line.replace(']'," ")
        line = line.replace('}'," ")
        line = line.replace(')'," ")
        
        line = [int(s) for s in line.split() if s.isdigit()]
        
        return line[0]

    def verify_Exe_Point(self, on, off, inPoint):
        
        addExePoint = False
        exeTrace = Event()

        if on == True and off == False:
            exeTrace.set_Event(inPoint)
            exeTrace.set_Line(self.extract_Line_Num(inPoint))
            self.eventManager.add_Event(exeTrace)
            addExePoint = True
            
        elif on == False and off == False:
            addExePoint = False
        else:
            addExePoint = False

        return addExePoint

    def exe_Point_Finder(self, trace): #This is really messy. Sorry. I will definitely need to refactor in the future
        symbolStack = []
        exePointList = []
        otherList = []
        currentSymbol = ''
        topSymbol = ''
        exe = ""
        exePoint = " "
        on = False
        off = False
        counter = 0
        
        for i in trace:
            currentSymbol = i
            exePoint = exePoint + currentSymbol
            if i == '{' or i == '[' or i == '(':
                symbolStack.append(i)
            elif i == '}' or i == ')' or i == ']':
                if self.is_empty(symbolStack) == False:
                    topSymbol = symbolStack.pop()
                    if i == '}' and topSymbol != '{':
                        continue
            elif i == ',':
                otherList.append(exePoint)
                if len(symbolStack) == 0:
                    for thing in otherList:
                        exe = exe + thing
                        
                    if "startTraceNow" in exe:
                        on = True
                        exe = ""
                        exePoint = ""
                        otherList = []
                    elif "endTraceNow" in exe:
                        off = True
                        return
                    else:
                        flag = self.verify_Exe_Point(on, off, exe)
                        if flag == True:
                            exe = ""
                            exePoint = ""
                            otherList = []
                        else:
                            on = False
                            exe = ""
                            exePoint = ""
                            otherList = []
                else:
                    exePoint = ""
            else:
                continue

def is_empty(structure):
    if structure:
        return False
    else:
        return True

def codeSplitter(code):

    studentCode = []

    code = code.split("startTraceNow();")
    newCode = code[1].split("endTraceNow();")

    executedCode = newCode[0]

    exeCodeList = executedCode.split("\\n")

    flag = False
    counter = 0 
    
    while flag != True:
        if exeCodeList[counter] == "" or exeCodeList[counter] == " ":
            flag = False
            counter = counter + 1
        elif exeCodeList[counter] != "":
            flag = True

    for x in range(counter, len(exeCodeList)):
        temp = exeCodeList[x]
        temp = "".join(temp.split())

        if is_empty(temp):
            studentCode.append("newline")
        else:
            studentCode.append(exeCodeList[x])

    return studentCode


def codeAnalyzer(code, firstTrace):
    
    codeToViz = []    
    codeToViz = codeSplitter(code)

    traceAnalyzer = TraceAnalyzer()

    execute = traceAnalyzer.handleEverything(codeToViz, firstTrace)

    


wholeTrace = wholeTrace[1:]
    
codeAnalyzer(userCode, wholeTrace)

    

